flatfs.go

```diff
@@ -404,6 +438,113 @@ func (fs *Datastore) walkTopLevel(path string, reschan chan query.Result) error
 	return nil
 }
 
+// calculateDiskUsage tries to read the diskUsageFile for a cached
+// diskUsage value, otherwise walks the datastore files.
+func (fs *Datastore) calculateDiskUsage() error {
+	// Try to obtain a previously stored value from disk
+	if persDu := fs.readDiskUsageFile(); persDu > 0 {
+		fs.diskUsageCh <- int64(persDu)
+		return nil
+	}
+
+	// Otherwise walk the datastore files
+	var du int64
+	err := filepath.Walk(fs.path,
+		func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			du += info.Size()
+			return nil
+		})
+
+	if err != nil {
+		log.Errorf("cannot calculate disk usage: %s", err)
+		return err
+	}
+
+	fs.diskUsageCh <- du
+	return nil
+}
+
+// diskUsageUpdater should run in a goroutine.
+// only call after calculateDiskUsage() has stored
+// a correct value in fs.diskUsage.
+func (fs *Datastore) diskUsageUpdater() {
+	var curDu int64
+
+	for duInc := range fs.diskUsageCh {
+		curDu += duInc // duInc might be < 0
+		atomic.StoreUint64(&fs.diskUsage, uint64(curDu))
+	}
+}
+
+func fileSize(path string) int64 {
```

Comment:
We could actually be *slightly* more accurate by adding in the size of the filename.

Reply:
you think? in my tests `du` on the disk matched the values we produce here. aren't filenames in ext* part of the directory entry, and the directory entry counts as fulll 4096 bytes block even if empty. So im not sure if adding filenames is double-counting some bytes..

Reply:
I was trying to account for the case where we add a bunch of files without re-stating the directories (in which case, they'll be much larger than 4096). However, that case may not really be worth considering. If it comes to it, we can always consider recomputing once in a while.

Reply:
ok, leaving this as it is. To my usecase, it doesn't really matter. We can improve it when we need to.

Reply:
Fair enough.